### PR TITLE
Karma Replacement

### PR DESCRIPTION
**Pull request by IanKeen**
- Better separation of concerns: each feature of karma is broken out 
- Moved command routing to new pattern match system
- Allowed for proper interaction based on public/private interaction (thanks @timvermeulen)
- Improvements to `Sugar` allowed for bug fixes to previous multi karma crashes
- Added "how much karma do I have?" command

### CODE REVIEW

File: Sources/Camille/Configs/Configs+CrossPostService.swift [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import Sugar
 
 extension Configs {
-    static let CrossPostService = CrossPostServiceConfig(
+    static let CrossPost = CrossPostServiceConfig(
```
Comment (aranasaurus) [inline Sources/Camille/Configs/Configs+CrossPostService.swift:4]: I haven't been super in-touch with the Swift 3 changes, and I realize the previous property was this as well, but it seems a little odd to me that these static properties are UpperCamelCased.

File: Sources/Camille/Karma/KarmaService+Config.swift [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import Models
+
+extension KarmaService {
+    struct Config {
+        let topUsersLimit: Int
+        let karmaAdjusters: [(adjuster: KarmaAdjustable, amount: Int)]
+        let textDistanceThreshold: Int
+        let allowedBufferCharacters: Set<Character>
+        let positiveMessage: (_ user: User, _ total: Int) -> [String]
+        let negativeMessage: (_ user: User, _ total: Int) -> [String]
```
Comment (aranasaurus) [inline Sources/Camille/Karma/KarmaService+Config.swift:11]: So much <3 for these returning arrays :D

**Review by aranasaurus — APPROVED** [review]
Looks good to me!

File: Sources/Camille/Configs/Configs+CrossPostService.swift (hunk 1 above)
Comment (ethankamcord) [reply Sources/Camille/Configs/Configs+CrossPostService.swift:4]: @aranasaurus I uppercase things in a Configs struct as well

**Comment by ethankamcord** [comment]
LGTM

File: Sources/Camille/Configs/Configs+CrossPostService.swift (hunk 1 above)
Comment (timvermeulen) [reply Sources/Camille/Configs/Configs+CrossPostService.swift:4]: I also find these properties a bit misleading... @IanKeen any idea if Linux Swift still doesn't like embedded types?

Comment (IanKeen) [reply Sources/Camille/Configs/Configs+CrossPostService.swift:4]: @aranasaurus I'm also kinda torn on these... tbh I don't really mind if these become `lowerCamelCase` but I guess I kinda think of these the same way as like a 'macro' in objc I guess? - but yeah happy to lowerCase these in a subsequent PR, especially if we also plan to move these `Config` structs to be nested types as @timvermeulen suggested. ( @timvermeulen I think they are ok to use now, so we can move to that pattern)

File: Sources/Camille/Karma/KarmaService+Config.swift (hunk 2 above)
Comment (IanKeen) [reply Sources/Camille/Karma/KarmaService+Config.swift:11]: 😄

File: Sources/Camille/Configs/Configs+CrossPostService.swift (hunk 1 above)
Comment (IanKeen) [reply Sources/Camille/Configs/Configs+CrossPostService.swift:4]: If everyone is cool with this style change being in another PR lemme know and I will merge this 👍

**Comment by aranasaurus** [comment]
Ship it!

**Pull request merged**